src/Sentry.Unity/Il2CppEventProcessor.cs

```diff
@@ -56,47 +54,163 @@ public void Process(Exception incomingException, SentryEvent sentryEvent)
-                    {
-                        codeFile = "GameAssembly.fallback";
-                    }
-                    debugImages.Add(new DebugImage
-                    {
-                        Type = _sentryUnityInfo.Platform,
+                _options.DiagnosticLogger?.LogDebug("NativeStackTrace Image: '{0}' (UUID: {1})", nativeStackTrace.ImageName, nativeStackTrace.ImageUuid);
 
-                        CodeFile = codeFile,
-                        DebugId = nativeStackTrace.ImageUuid,
-                        ImageAddress = $"0x{imageAddress:X8}",
-                    });
-
-                    didAddImage = true;
-                }
+                // Unity by definition only builds a single library which we add once to our list of debug images.
+                // We use this when we encounter stack frames with relative addresses.
+                // We want to use an address that is definitely outside of any address range used by real libraries.
+                // Canonical addresses on x64 leave a gap in the middle of the address space, which is unused.
+                // This is a range of addresses that we should be able to safely use.
+                // See https://en.wikipedia.org/wiki/X86-64#Virtual_address_space_details
+                var mainLibOffset = (ulong)1 << 63;
+                DebugImage? mainLibImage = null;
 
+                // TODO do we really want to continue if these two don't match?
+                //      Wouldn't it cause invalid frame info?
                 var nativeLen = nativeStackTrace.Frames.Length;
                 var len = Math.Min(sentryStacktrace.Frames.Count, nativeLen);
                 for (var i = 0; i < len; i++)
                 {
-                    // The sentry stack trace is sorted parent (caller) to child (callee),
+                    // The sentry stack trace is sorted parent->child (caller->callee),
                     // whereas the native stack trace is sorted from callee to caller.
                     var frame = sentryStacktrace.Frames[i];
                     var nativeFrame = nativeStackTrace.Frames[nativeLen - 1 - i];
+                    var nativeImageUUID = NormalizeUUID(nativeStackTrace.ImageUuid);
 
                     // The instructions in the stack trace generally have "return addresses"
                     // in them. But for symbolication, we want to symbolicate the address of
                     // the "call instruction", which in almost all cases happens to be
                     // the instruction right in front of the return address.
                     // A good heuristic to use in that case is to just subtract 1.
-                    var instructionAddr = imageAddress + nativeFrame.ToInt64() - 1;
-                    frame.InstructionAddress = $"0x{instructionAddr:X8}";
+                    var instructionAddress = (ulong)nativeFrame.ToInt64() - 1;
+                    var image = FindDebugImageContainingAddress(instructionAddress);
```

Comment:
this might give you some false positives :-(

In one of the examples I have seen that one of the images had an addr at `0x40_000`, and there were some GameAssembly-relative addresses that just happen to be inside of that :-(

Reply:
The issue is, now all the code only matches through the package-name code branch... At least with the error examples we have in the sample. Any ideas how to create stack trace with items that would match other debug images? Otherwise it seems kinda pointless to even post them.

Reply:
BTW I've tried adding an example with a callback through native code, but that is still part of GameAssembly. Other option I can think of is using an external library, e.g. sentry, with one of its callbacks, e.g. before_send